File: 명진/programmers/보석 쇼핑.py
def solution(gems):
    al, ar = 0, len(gems)-1
    numGems = len(set(gems))
    gemCounts = {gems[0]:1}
    l = r = 0
    while l < len(gems) and r < len(gems):
        if len(gemCounts) == numGems:
            if r - l < ar - al:
                ar, al = r, l
            gemCounts[gems[l]] -= 1
            if gemCounts[gems[l]] == 0:
                del gemCounts[gems[l]]
            l += 1
            continue
        r+=1
        if r == len(gems):
            break
        gemCounts[gems[r]] = gemCounts.get(gems[r], 0) + 1
    return [al+1, ar+1]

File: 규진/프로그래머스/week16/보석_쇼핑.cpp
#include <string>
#include <vector>
#include <set>
#include <map>
#include <iostream>

using namespace std;

vector<int> solution(vector<string> gems) {
    vector<int> answer;
    set<string> s;
    map<string, int> m;
    int start = 0, end = 0;
    
    for (int i = 0; i < gems.size(); i++)
        s.insert(gems[i]);
    
    for (auto i : gems) {
        m[i]++;
        if (m.size() == s.size()) {
            break;
        }
        end++;
    }
    
    // for (auto iter = m.begin(); iter != m.end(); iter++) {
    //     cout << iter->first << " : " << iter->second << endl;
    // }
    
    int m_size = end;

    while (true) {
        int cnt = 0;
        auto key = gems[start];
        m[key]--;
        for (auto iter = m.begin(); iter != m.end(); iter++) {
            if (iter->second > 0)
                cnt++;
        }
        
        if (cnt != s.size()) {
            break;
        } else {
            m_size = cnt;
            start++;
        }
    }
    start = start + 1;
    end = end + 1;
    
    answer.push_back(start);
    answer.push_back(end);
    
    return answer;
}